h2o-genmodel/build.gradle

```diff
@@ -27,6 +27,8 @@ dependencies {
   testCompile "junit:junit:${junitVersion}"
   compile "net.sf.opencsv:opencsv:2.3"
   compile "com.google.code.gson:gson:2.6.2"
+  compile "guru.nidi:graphviz-java:0.9.0"
+  compile "org.slf4j:slf4j-log4j12:1.7.10"
```

Comment:
This dependency could be acceptable as test-scoped. We don't want SLF4J implementations to be introduced to the classpath by H2O - Genmodel uses system output directly. SLF4J dependencies are even specifically excluded few lines below.

Reply:
1. slf4j can be removed from dependencies. However, there will be these logs written into console on each execution: 

`Zuzanas-MacBook-Pro:h2oexp zuzanaolajcova$  java -cp h2o.jar hex.genmodel.tools.PrintMojo --tree 0 -i test_h2o.csv.zip -o test_h2o.csv.gv -f 20 -d 3 --direct mynewpng.png 
log4j:WARN No appenders could be found for logger (guru.nidi.graphviz.service.CommandLineExecutor).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.` 

But maybe there is a way how to suppress it.

2. slf4j cannot be excluded from graphviz, since it is used by its classes and it would lead to java.lang.NoClassDefFoundError: org/slf4j/LoggerFactory

Reply:
There are multiple options we have now. 

## Providing minimal SLF4J implementation
The first option requires making ourselves comfortable with slf4j being present on the classpath. GenModel is aimed to be present on the classpath of other projects directly and we need to make sure any SLF4J implementations and most importantly configurations do not interfere, override or change the behavior for our customers.

It should also be done in the minimal way possible (we keep an eye for `genmodel` module's dependencies to be as slim as possible). This means preventing any transitive dependencies on SLF4J API and only supplying a minimal implementation. LOG4J previously present has around 500k, it's only true alternative Logback is about the same size. Using the `SimpleLogger` provided by SLF4J directly introduces a 15K dependency:

```groovy
  compile ("guru.nidi:graphviz-java:0.9.0"){
    exclude group: 'org.slf4j'
  }
  compile group: 'org.slf4j', name: 'slf4j-simple', version: '1.7.27'
```

Please note any transitive slf4j dependencies are excluded, thus no API will be directly present (the scope of the API could actually be compileOnly or "provided" in Maven terms in the graphviz-java project). It is perfectly fine with the `'slf4j-simple` implementation.

There is only one disadvantage - `SimpleLogger` by default prints into System.err, not stdout. This can be further configured, if required. However, I am almost 100% sure this is OK, as this is mostly used as a command line tool and the output for debugging will still be there somewhere, even if the error output is not shown to the user directly in some cases.

## Defaulting to NO-OP

Looks nasty, I'd not do it. There is still some transitive dependency required (otherwise we'll get `NoClassDefFoundError`), so this really solves nothing for us.

## Fork graphviz-java

Forking the repository and substituting `Log.*` with direct usage if `System.out` would be possible and it wouldn't even be that much work, considering the size of the project. However, the downsides are obvious. No easy updates, which means no bugfixes for free. Plus the cost of custom maintenance.

## Outcome
Personally, I'd choose the way of **providing minimal SLF4J implementation**. WDYT @michalkurka , is it safe to include minimal SLF4J implementation with size of 15k, wihch only contains the bare minimum ? The source code that will get included can be seen here: https://github.com/qos-ch/slf4j/tree/master/slf4j-simple/src/main/java/org/slf4j/simple

This way, the classpath content regarding SLF4J would be controlled by us.

What is your opinion @koniecsveta ?

Reply:
I think that if there is logging already implemented in graphviz-java, we should use it (logs can be helpful). I would go for simple logger which would store graphviz-java log-output with some acceptable backup strategy (size/date). The question is what is acceptable in this case? If logging is not acceptable and the tool we want to integrate has it as a prerequisite, then we should maybe try different tool (or create own by fork but that comes with mentioned cons).